### PR TITLE
AKS credentials for ACR

### PR DESCRIPTION
**Pull request by alanjcastonguay**
Azure Kubernetes Service (AKS) nodes' kubelet can use the cluster's
service principal to fetch images from Azure Container Registry (ACR).

If azure.json is projected into the Flux pod, consume the service
principal credentials when authenticating to azure container registry.

Note that this method may eventually be deprecated, possibly replaced
by Managed Identity + OAuth, similar to the GCP implementation.
See https://github.com/kubernetes/kubernetes/issues/58034

Prompted by [discussion](https://weave-community.slack.com/archives/C4U5ATZ9S/p1547928858828100) with @squaremo on slack

### CODE REVIEW

File: registry/azure.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    AADClientSecret string `json:"aadClientSecret"`
+}
+
+// Fetch Azure Active Directory clientid/secret pair from azure.json, usable for container registry authentication.
+//
+// Note: azure.json is populated by AKS/AKS-Engine script kubernetesconfigs.sh. The file is then passed to kubelet via
+// --azure-container-registry-config=/etc/kubernetes/azure.json, parsed by kubernetes/kubernetes' azure_credentials.go
+// https://github.com/kubernetes/kubernetes/issues/58034 seeks to deprecate this kubelet command-line argument, possibly
+// replacing it with managed identity for the Node VMs. See https://github.com/Azure/acr/blob/master/docs/AAD-OAuth.md
+func GetAzureCloudConfigAADToken(host string) (creds, error) {
```
Comment (squaremo) [inline registry/azure.go]: I don't think this or `HostIsAzureContainerRegistry` need to be exported -- they are just used in the tests and in `credsFor`, both in the same package.

Comment (alanjcastonguay) [reply registry/azure.go]: I was matching the pattern of `GetGCPOauthToken` being exported - it's also only used in those two places in the `registry` package. I wasn't sure if that was intentional. I can make this private anyway.

Comment (squaremo) [reply registry/azure.go]: Fair point, following the apparent convention ought not earn you demerits :-)

**Comment by 4c74356b41** [comment]
I can confirm this is working by adding these changes to the deploy/flux:
```
        volumeMounts:
        xxx
        - mountPath: /etc/kubernetes/azure.json
          name: acr-credentials
        xxx
      volumes:
      xxx
      - hostPath:
          path: /etc/kubernetes/azure.json
          type: ""
        name: acr-credentials
```

**Comment by squaremo** [comment]
> I can confirm this is working

Splendid, thank you very much for trying it out @4c74356b41

**Comment by alanjcastonguay** [comment]
Awesome!

Let me fix the formatting and docs....

**Comment by timwebster9** [comment]
ACR user here...

Just so I understand, does this PR remove the need to provide `imagePullSecrets` along with your k8s resources in order to enable the registry scanning to work?

Oddly enough, that method works for me, unlike what people are reporting in #1396.  But this method looks nicer :-)

Also, will this help with ACR Helm authentication at all?

**Comment by 4c74356b41** [comment]
In my experiments using imagePullSecrets doesnt work if you grant AKS permissions to ACR, so you have to strip AKS permissions on ACR and after that it will work with imagePullSecrets.

Not sure what do you mean about helm and ACR. helm is a client side tool

**Comment by timwebster9** [comment]
@4c74356b41 

The SP my cluster is using has AcrPull role assignment, and I'm using a different SP (also with AcrPull) for the credentials used in the `imagePullSecrets`.  It seems to work.

Re: the ACR Helm thing - sorry I wasn't clear.  ACR can also host Helm charts (i.e. it's a Helm repository), but since it's private it requires authentication.  

https://docs.microsoft.com/en-us/azure/container-registry/container-registry-helm-repos

I'm guessing nobody has tried this or looked at it.  I tried briefly using the method described [here](https://github.com/weaveworks/flux/blob/master/site/helm-integration.md#authentication-for-helm-repos), but it didn't work.  It's not an urgent requirement for us at the moment so I haven't made time to investigate further.  The authentication method shown in the docs uses an OAuth token which expires after an hour, so not sure this will work.

However when using ACR as a container registry, you can authenticate directly using a service principal (i.e. with `docker login`), so I was wondering if this would work with the Helm repo authentication somehow.

**Comment by 4c74356b41** [comment]
I'm well aware of that, what I dont understand, with flux all your templates are in the repo already, doesnt matter you have them in a private repo as well? also, if you have acr pull for AKS sp it wont even use imagePullSecrets

**Comment by timwebster9** [comment]
> I'm well aware of that, what I dont understand, with flux all your templates are in the repo already, doesnt matter you have them in a private repo as well? also, if you have acr pull for AKS sp it wont even use imagePullSecrets

The reason I needed `imagePullSecrets` was not for AKS to pull images, but for Flux to poll/scan the repositories.  Unless I'm mistaken, Flux doesn't use the AKS SP credentials to do this. 

So back to the my original question: my guess is that this PR will provide those same credentials (via the host mount) to Flux, so we don't need the `imagePullSecrets`.  Does this sound right?

We currently have some internal Helm charts published to ACR, so I wanted to see if Flux could install those.  However we don't need to do this, rather we can just install them from Github instead.

I'm not sure what you mean by `your templates are in the repo already`.  What repo?  The Helm charts are in separate repos, not in the flux one.  The Flux repo only has the `HelmRelease` resources that reference these charts.

**Comment by 4c74356b41** [comment]
oh, ok. I didnt knew you can do that. I thought it needed charts in the same repo.

indeed flux doesnt use aks sp credentials (this is why this PR exists), but what I'm telling you is that the solution worked for you because you didnt have ACRpull for the AKS sp, if you did - it wouldnt.

**Comment by timwebster9** [comment]
> oh, ok. I didnt knew you can do that. I thought it needed charts in the same repo.
> 
> indeed flux doesnt use aks sp credentials (this is why this PR exists), but what I'm telling you is that the solution worked for you because you didnt have ACRpull for the AKS sp, if you did - it wouldnt.

I have ACR pull on both the AKS SP and the one I use for `imagePullSecrets' and everything works fine.  I guess what you're saying is you need to use `imagePullSecrets` both for polling the registry *and* pulling images?

Before I added `imagePullSecrets` to my Flux resources AKS could pull images fine using the usual SP role assignment on ACR.  I only had to add it for the registry scan bit.

**Comment by alanjcastonguay** [comment]
@squaremo corrected whitespace (via gofmt) in 3cb2179, un-Exported two methods in df5ca66, added docs to faq.md in cf30277, and plumbed hostPath mount into helm chart (disabled by default, always readOnly) in 0ce0668. I think that's all the requested changes.

**Comment by 4c74356b41** [comment]
@alanjcastonguay cool, didnt think about readOnly, makes a lot of sense.
@timwebster9 ok, i guess you are special :)

**Comment by squaremo** [comment]
@alanjcastonguay This is a thorough job, thank you very much!

File: chart/flux/templates/deployment.yaml [hunk 2]
```diff
@@ -45,6 +45,12 @@ spec:
       - name: git-keygen
         emptyDir:
           medium: Memory
+      {{- if .Values.registry.acr.enabled }}
+      - name: acr-credentials
+        hostPath:
+          path: "{{ .Values.registry.acr.mountPath }}"
```
Comment (hiddeco) [inline chart/flux/templates/deployment.yaml]:
```suggestion
          path: "{{ .Values.registry.acr.hostPath }}"
```

as per docs and `values.yaml` example.

Comment (alanjcastonguay) [reply chart/flux/templates/deployment.yaml]: Thanks for the catch! Looks like this is the only place I made that mistake.

**Pull request merged**